stan/math/prim/prob/wiener5_lpdf.hpp

```diff
@@ -0,0 +1,428 @@
+    } else {
+      u_eps = fmin(-3.0,
+                   (log(8.0) - log(27.0) + LOG_PI + 4.0 * log_y_asq + two_es));
+    }
+    double arg_mult = (FunTypeEnum == FunType::Density) ? 1 : 3;
+    arg = -arg_mult * y_asq * (u_eps - sqrt(-2.0 * u_eps - 2.0));
+
+    double K2 = (arg > 0) ? 0.5 * (sqrt(arg) - w) : K1;
+    double kss = ceil(fmax(K1, K2));
+
+    static const double PISQ = square(pi());  // pi*pi
+    if (FunTypeEnum == FunType::Density) {
+      K1 = 1.0 / (pi() * sqrt(y_asq));
+      double two_log_piy = -2.0 * (LOG_PI + log_y_asq + es);
+      K2 = (two_log_piy >= 0) ? sqrt(two_log_piy / (PISQ * y_asq)) : 0.0;
+    } else {
+      K1_mult = (FunTypeEnum == FunType::GradT) ? 2 : 3;
+      K1 = sqrt(K1_mult / y_asq) / pi();
+      double u_eps_arg
+          = (FunTypeEnum == FunType::GradT)
+                ? log(3.0) - log(5.0) + LOG_PI + 2.0 * log_y_asq + es
+                : log(4.0) - log(9.0) + 2.0 * LOG_PI + 3.0 * log_y_asq + two_es;
+      u_eps = fmin(-1, u_eps_arg);
+      arg_mult = (FunTypeEnum == FunType::GradT) ? (2.0 / PISQ / y_asq) : 1;
+      arg = -arg_mult * (u_eps - sqrt(-2.0 * u_eps - 2.0));
+      K2 = (FunTypeEnum == FunType::GradT)
+               ? (arg > 0) ? sqrt(arg) : K1
+               : (arg > 0) ? sqrt(arg / y_asq) / pi() : K1;
+    }
+
+    double kll = ceil(fmax(K1, K2));
+
+    double ld_err = 0;
+    if (FunTypeEnum != FunType::Density) {
+      if (FunTypeEnum == FunType::GradT) {
+        ld_err = log(max(fabs(ans0 - 1.5 / y), fabs(ans0)));
+      } else if (FunTypeEnum == FunType::GradA) {
+        ld_err = log(max(fabs(ans0 + 1.0 / a), fabs(ans0 - 2.0 / a)));
+      } else {
+        ld_err = log(fabs(ans0));
+      }
+      ld = wiener5_helper<FunType::Density>(y, a, vn, wn, sv, err - ld_err);
+    }
+
+    double erg = NEGATIVE_INFTY;
+    int newsign = 1;
+
+    bool kss_comp
+        = (FunTypeEnum == FunType::Density) ? 2 * kss <= kll : 2 * kss < kll;
+    double scaling = kss_comp ? inv(2.0 * y_asq) : y_asq / 2.0;
+    if (kss_comp) {
+      double mult = (FunTypeEnum == FunType::Density) ? 1 : 3;
+      double offset = (FunTypeEnum == FunType::GradW) ? y_asq : 0;
+      double sqrt_offset = sqrt(offset);
+      for (size_t k = static_cast<size_t>(kss); k >= 0; k--) {
+        double wp2k = w + 2.0 * k;
+        double wm2k = w - 2.0 * k;
+        int wp2k_sign = (wp2k > sqrt_offset) ? 1 : -1;
+        int wm2k_sign = (wm2k > sqrt_offset) ? 1 : -1;
+        double wp2k_quant = log(wp2k_sign * (wp2k - offset))
+                            - (square(wp2k) - offset) * scaling;
+        double wm2k_quant = log(wm2k_sign * (wm2k - offset))
+                            - (square(wm2k) - offset) * scaling;
+        std::forward_as_tuple(erg, newsign)
+            = log_sum_exp_signed(erg, newsign, mult * wp2k_quant, wp2k_sign);
+        std::forward_as_tuple(erg, newsign) = log_sum_exp_signed(
+            erg, newsign, mult * wm2k_quant, -1 * wm2k_sign);
```

Comment:
It is sort of code smell that the things we are assigning to are also inputs to the function

Reply:
Out of scope for this PR, would require changes to the `log_sum_exp_signed` function